### PR TITLE
Update payment module for CS-Cart 4.3.x

### PR DESCRIPTION
**Pull request by dzungtran**
Hi @davex69 and @ivan-kovalyov,
I've updated this module for CS-Cart 4.3.x . Please review it.

### CODE REVIEW

File: src/app/controllers/frontend/paymentwall.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+	foreach ($params as $k=>$v) {
+		$str .= "$k=$v";
+	}
+	$str .= $secret;
+
+	return md5($str);
+}
+
+?>
```
Comment (sevastyanovio) [inline src/app/controllers/frontend/paymentwall.php]: Use [Paymentwall lib](https://github.com/paymentwall/paymentwall-php)

File: src/js/jquery.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/*! jQuery v1.7.1 jquery.com | jquery.org/license */
```
Comment (sevastyanovio) [inline src/js/jquery.js]: Remove jQuery, use VanillaJS

File: src/app/payments/paymentwall_window.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
+<?php
```
Comment (sevastyanovio) [inline src/app/payments/paymentwall_window.php]: Is it possible to separate PHP and HTML code?

File: src/app/controllers/frontend/paymentwall.php (hunk 1 above)
Comment (dzungtran) [inline src/app/controllers/frontend/paymentwall.php]: Okay, i can change

File: src/js/jquery.js (hunk 2 above)
Comment (dzungtran) [inline src/js/jquery.js]: okay, no problem :)

File: src/app/controllers/frontend/paymentwall.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    $processor_params = db_get_field("SELECT processor_params FROM ?:payments WHERE payment_id = ?s", $payment_id);
+    $payment_configs = unserialize($processor_params);
+
+    Paymentwall_Config::getInstance()->set(array(
+        'api_type' => Paymentwall_Config::API_GOODS,
+        'public_key' => isset($payment_configs['key']) ? $payment_configs['key'] : null,
+        'private_key' => isset($payment_configs['secret']) ? $payment_configs['secret'] : null,
+    ));
+
```
Comment (sevastyanovio) [inline src/app/controllers/frontend/paymentwall.php]: move to external method

File: src/app/payments/paymentwall_window.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+if (!defined('BOOTSTRAP')) { die('Access denied');}
+
+use Tygh\Registry;
+require_once Registry::get('config.dir.lib') . DS . 'other' . DS . 'paymentwall-php' . DS . 'lib' . DS . 'paymentwall.php';
+
+Paymentwall_Config::getInstance()->set(array(
+    'api_type' => Paymentwall_Config::API_GOODS,
+    'public_key' => $processor_data['processor_params']['key'],
+    'private_key' => $processor_data['processor_params']['secret']
+));
```
Comment (sevastyanovio) [inline src/app/payments/paymentwall_window.php]: move to external method

File: src/app/addons/paymentwall/addon.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<addon scheme="3.0" edition_type="ROOT,ULT:VENDOR">
+    <id>paymentwall</id>
+    <version>1.0</version>
```
Comment (davex69) [inline src/app/addons/paymentwall/addon.xml]: semver.org

File: src/app/addons/paymentwall/config.php [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+if (!defined('BOOTSTRAP')) { die('Access denied'); }
```
Comment (davex69) [inline src/app/addons/paymentwall/config.php]: use PSR-2

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        if (count($orderInfo['products']) > 0) {
+            $products[] = new Paymentwall_Product($orderId, $realPrice, $currencyCode, 'Order #' . $orderId);
+        }
+
+        $widget = new Paymentwall_Widget($uid, $widgetCode, $products, array(
+            'email' => $orderInfo['email'],
+            'payment_id' => $orderInfo['payment_id'],
+            'integration_module' => 'cs_cart',
+            'test_mode' => $testMode,
+            'ref' => rand(99, 999)
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: ref in not required
https://www.paymentwall.com/en/documentation/Digital-Goods-API/710#default_widget_call_parameters

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+if ($mode == 'pingback') {
+
+    define('CREDIT_TYPE_CHARGEBACK', 2);
+    unset($_GET['dispatch']);
+
+    $orderId = isset($_GET['goodsid']) ? $_GET['goodsid'] : null;
+    $paymentId = isset($_GET['payment_id']) ? $_GET['payment_id'] : null;
+    $type = isset($_GET['type']) ? $_GET['type'] : null;
+    $result = false;
+
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: move to external function getPingbackParams()

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    Tygh::$app['view']->assign('matchOrder', $matchOrder);
+    Tygh::$app['view']->assign('orderId', $orderId);
+    Tygh::$app['view']->assign('baseUrl', $baseUrl);
+    Tygh::$app['view']->assign('iframe', $iframe);
+}
+
+if ($mode == 'pingback') {
+
+    define('CREDIT_TYPE_CHARGEBACK', 2);
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: define constants at the top of file

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+    $orderId = isset($_GET['goodsid']) ? $_GET['goodsid'] : null;
+    $paymentId = isset($_GET['payment_id']) ? $_GET['payment_id'] : null;
+    $type = isset($_GET['type']) ? $_GET['type'] : null;
+    $result = false;
+
+    $configs = getPaymentConfigs($paymentId);
+    $result = handlePingback($configs, $orderId, $type);
+
+    if ($result) die('OK');
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: set 'OK' as constant DEFAULT_PINGBACK_RESPONSE

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            db_query('UPDATE ?:orders SET ?u WHERE order_id = ?i', $data, $orderId);
+
+            // Take membership from user
+            // This is optional, but we recommend this type of crediting to be implemented as well
+            //$this->model_checkout_order->update($goodsId, $this->config->get('cancel_status'));
+        } else {
+            $data = array(
+                'status' => 'P'
+            );
+            db_query('UPDATE ?:orders SET ?u WHERE order_id = ?i', $data, $orderId);
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: DRY

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    ));
+
+    $pingback = new Paymentwall_Pingback($params, $_SERVER['REMOTE_ADDR']);
+
+    if ($pingback->validate()) {
+        if ($type == CREDIT_TYPE_CHARGEBACK) {
+            $data = array(
+                'status' => 'I'
+            );
+            db_query('UPDATE ?:orders SET ?u WHERE order_id = ?i', $data, $orderId);
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: DRY

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+    $params = array_merge($_GET, array(
+        'sign_version' => Paymentwall_Signature_Abstract::VERSION_THREE
+    ));
+
+    $pingback = new Paymentwall_Pingback($params, $_SERVER['REMOTE_ADDR']);
+
+    if ($pingback->validate()) {
+        if ($type == CREDIT_TYPE_CHARGEBACK) {
+            $data = array(
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: move to external function getStatusData(), DRY

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    exit;
+}
+
+function handlePingback($configs, $orderId, $type)
+{
+    Paymentwall_Config::getInstance()->set(array(
+        'api_type' => Paymentwall_Config::API_GOODS,
+        'public_key' => isset($configs['key']) ? $configs['key'] : null,
+        'private_key' => isset($configs['secret']) ? $configs['secret'] : null,
+    ));
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: move to external function initPaymentwallSdk()

And init there sdk for Widget Call and Pingback

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            );
+            db_query('UPDATE ?:orders SET ?u WHERE order_id = ?i', $data, $orderId);
+
+            // Take membership from user
+            // This is optional, but we recommend this type of crediting to be implemented as well
+            //$this->model_checkout_order->update($goodsId, $this->config->get('cancel_status'));
+        } else {
+            $data = array(
+                'status' => 'P'
+            );
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: move to external function getStatusData(), DRY

File: src/app/addons/paymentwall/init.php [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<?php
+
+if (!defined('BOOTSTRAP')) { die('Access denied'); }
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/init.php]: move to external function, DRY

```
if (!defined('BOOTSTRAP')) {
    die('Access denied');
}
```

File: src/app/addons/paymentwall/payments/paymentwall.php [hunk 18]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+if (!defined('BOOTSTRAP')) {
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/payments/paymentwall.php]: move to external function, DRY

File: src/app/addons/paymentwall/payments/paymentwall.php [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    die('Access denied');
+}
+
+use Tygh\Registry;
+
+require_once Registry::get('config.dir.lib') . DS . 'other' . DS . 'paymentwall-php' . DS . 'lib' . DS . 'paymentwall.php';
+
+$coefficient = db_get_field("SELECT coefficient FROM ?:currencies WHERE currency_code = ?s", $order_info['secondary_currency']);
+$realPrice = $order_info['total'] / $coefficient;
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/payments/paymentwall.php]: function getRealPrice($total)

File: src/app/addons/paymentwall/controllers/common/paymentwall.php (hunk 8 above)
Comment (dzungtran) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: U sure?
I see some code in pw lib: class Paymentwall_Pingback and isSignatureValid method

``` php
} else if ($this->getApiType() == Paymentwall_Config::API_GOODS) {
    $signatureParams = array('uid', 'goodsid', 'slength', 'speriod', 'type', 'ref');
}
```

File: src/app/addons/paymentwall/payments/paymentwall.php (hunk 18 above)
Comment (dzungtran) [inline src/app/addons/paymentwall/payments/paymentwall.php]: I think it not matter, it still follows concepts of CMS

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 20]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+if (!defined('AREA')) { die('Access denied');}
+
+use Tygh\Registry;
+
+if ($mode == 'ajax') {
+    $status = db_get_field("SELECT status FROM ?:orders WHERE order_id = ?i", $_REQUEST['order_id']);
+    if ($status == "P") {
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: function getStatus($mode)

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+if ($mode == 'frame') {
+    Tygh::$app['view']->assign('orderId', $_REQUEST['order_id']);
+}
+
+// Payment widget
+if ($mode == 'payment') {
+
+    $rid = $_REQUEST['order_id'];
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php:23]: move to prepareData() function

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    Tygh::$app['view']->assign('orderId', $orderId);
+    Tygh::$app['view']->assign('baseUrl', $baseUrl);
+    Tygh::$app['view']->assign('iframe', $iframe);
+}
+
+if ($mode == 'pingback') {
+
+    unset($_GET['dispatch']);
+
+    $orderId = isset($_GET['goodsid']) ? $_GET['goodsid'] : null;
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: function prepareData($mode)

File: src/app/addons/paymentwall/func.php [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+        'public_key' => $projectKey,
+        'private_key' => $secretKey
+    ));
+}
+
+function fn_paymentwall_generateWidget($orderInfo,$paymentInfo, $extraAttr = array())
+{
+    fn_paymentwall_initPaymentwallSdk($paymentInfo['key'], $paymentInfo['secret']);
+
+    $uid = empty($orderInfo['user_id']) ? $orderInfo['ip_address'] : $orderInfo['user_id'];
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/func.php]: function prepareData()

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+            break;
+        default:
+            break;
+    }
+}
+
+function orderStatusProcessor($status)
+{
+    switch ($status) {
+        case 'P':
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: 'P' used many times, better - create const for this status and use this const

same for other statues

File: src/app/addons/paymentwall/func.php [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+    $products = array();
+    $realPrice = fn_paymentwall_getRealPrice($orderInfo);
+
+    if (count($orderInfo['products']) > 0) {
+        $products[] = new Paymentwall_Product($orderInfo['order_id'], $realPrice, $orderInfo['secondary_currency'], 'Order #' . $orderInfo['order_id']);
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/func.php]: move to function prepareProducts()

File: src/app/addons/paymentwall/func.php [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    $widget = new Paymentwall_Widget($uid, $paymentInfo['widget_type'], $products, array(
+        'email' => $orderInfo['email'],
+        'payment_id' => $orderInfo['payment_id'],
+        'integration_module' => 'cs_cart',
+        'test_mode' => $paymentInfo['test_mode'],
+        'ref' => rand(99, 999),
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/func.php]: remove

File: src/app/addons/paymentwall/payments/paymentwall.php [hunk 27]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+if (!defined('AREA')) { die('Access denied'); }
+
+$orderId = $order_info['order_id'];
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/payments/paymentwall.php]: instead $orderId  use $order_info['order_id'];

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+        'baseUrl' => $baseUrl,
+        'iframe' => $iframe
+    ));
+}
+
+function modePingback()
+{
+    unset($_GET['dispatch']);
+
+    $orderId = isset($_GET['goodsid']) ? $_GET['goodsid'] : null;
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: remove, use $pingback->getProduct()->getId();

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+    $paymentId = isset($_GET['payment_id']) ? $_GET['payment_id'] : null;
+    $type = isset($_GET['type']) ? $_GET['type'] : null;
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: remove, use $pingback->getProduct()->getType();

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+
+// Payment widget
+function modePayment()
+{
+    $rid = $_REQUEST['order_id'];
+    $sid = $_SESSION['pw_order_id'];
+    $iframe = '';
+    $baseUrl = fn_url();
+    $matchOrder = ($rid == $sid);
+    $orderId = $sid;
```
Comment (sevastyanovio) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: remove

File: src/app/addons/paymentwall/controllers/common/paymentwall.php [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+    $orderId = isset($_GET['goodsid']) ? $_GET['goodsid'] : null;
+    $paymentId = isset($_GET['payment_id']) ? $_GET['payment_id'] : null;
+    $type = isset($_GET['type']) ? $_GET['type'] : null;
+
+    $configs = fn_paymentwall_getPaymentConfigs($paymentId);
+    $result = fn_paymentwall_handlePingback($configs, $orderId, $type);
+
+    if ($result) die(PW_DEFAULT_PINGBACK_RESPONSE);
+
+    exit;
```
Comment (davex69) [inline src/app/addons/paymentwall/controllers/common/paymentwall.php]: remove

**Pull request merged**